infrastructure/src/main/java/org/corfudb/infrastructure/protocol/NettyServerRouter.java

```diff
@@ -0,0 +1,144 @@
+import org.corfudb.common.protocol.proto.CorfuProtocol.MessageType;
+import org.corfudb.common.protocol.proto.CorfuProtocol.Request;
+import org.corfudb.common.protocol.proto.CorfuProtocol.Header;
+import org.corfudb.infrastructure.ServerContext;
+import org.corfudb.runtime.view.Layout;
+
+import java.util.*;
+
+@Slf4j
+@ChannelHandler.Sharable
+public class NettyServerRouter extends ChannelInboundHandlerAdapter implements IServerRouter {
+
+    /**
+     * This map stores the mapping from message types to server handler.
+     */
+    private final Map<MessageType, AbstractServer> handlerMap;
+
+    /**
+     * This node's server context.
+     */
+    private final ServerContext serverContext;
+
+    /**
+     * The epoch of this router. This is managed by the base server implementation.
+     */
+    @Getter
+    @Setter
+    volatile long serverEpoch;
+
+    /** The {@link AbstractServer}s this {@link NettyServerRouter} routes messages for. */
+    private final ImmutableList<AbstractServer> servers;
+
+    public NettyServerRouter(ImmutableList<AbstractServer> servers, ServerContext serverContext) {
+        this.serverContext = serverContext;
+        this.serverEpoch = serverContext.getServerEpoch();
+        this.servers = servers;
+        handlerMap = new EnumMap<>(MessageType.class);
+
+        servers.forEach(server -> {
+            Set<MessageType> handledTypes = server.getHandler().getHandledTypes();
+            handledTypes.forEach(handledType -> handlerMap.put(handledType, server));
+        });
+    }
+
+    @Override
+    public Optional<Layout> getCurrentLayout() {
+        return Optional.ofNullable(serverContext.getCurrentLayout());
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @deprecated This operation is no longer supported. The router will only route messages for
+     * servers provided at construction time.
+     */
+    @Override
+    @Deprecated
+    public void addServer(AbstractServer server) {
+        //TODO(Zach): Method still needed?
+        throw new UnsupportedOperationException("No longer supported");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<AbstractServer> getServers() {
+        return servers;
+    }
+
+    @Override
+    public void setServerContext(ServerContext serverContext) {
+        //TODO(Zach): Method still needed?
+        throw new UnsupportedOperationException("The operation is not supported.");
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        log.info("channelActive: Incoming connection established from: {}.",
+                ctx.channel().remoteAddress());
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ByteBuf msgBuf = (ByteBuf) msg;
+
+        // Temporary -- Check Corfu msg marker: 0x1 indicates legacy while 0x2 indicates new
+        if (msgBuf.getByte(msgBuf.readerIndex()) == 0x1) {
```

Comment:
Can you make these static constants (i.e. 0x1 and 0x2)

Reply:
@zfrenette  Did you delete your comment?

I got an email notification but don't see it on the PR.

> Do you mean throw an exception (instead of passing it along the pipeline) if the message isn't marked with PROTO_CORFU_MSG_MARK?


Yes, I think its better to throw an exception instead of the propagate the error further.